### PR TITLE
Add VI bindings

### PR DESCRIPTION
**Pull request by Ambrevar**

### CODE REVIEW

**Comment by alexanderbarbosa** [comment]
ITS HAPPENING :D

**Comment by Ambrevar** [comment]
The following features are missing and should be implemented as soon as possible for an ideal experience.

- Allow setting vi-normal-mode by default.
- Declare known keymap scheme names (`:emacs`, etc.) to avoid typos.
- Define `keymap` structure (so that we have KEYMAP-P).
- Disable VI mode properly.
- Auto-switch to insert mode when clicking on HTML text fields
- Translate "shift + key" to "KEY", e.g. "s-:" should translate to ":" if unbound.
- Add support for override maps.

## Implementation details

If you look at the `vi.lisp` files, you'll see it's some 30 trivial lines of code for basic VI support!

I thought quite a lot about it and decided to first make Next as general as possible regarding the key facilities so that any key binding scheme could be implemented with the least amount of effort as possible.

In fact, the beauty of Next is that we don't depend on either the history of Emacs nor the limitations of WebExtensions or Firefox.  We have full control here, so let's design something neat from the ground up.

- Buffers

`buffer` is a Common Lisp class and defines a whole bunch of slots.
One of those slots, `modes`, holds the list of active modes for this buffer.
The first mode has highest priority.

Another slot is the `current-key-scheme` (see below).

- Modes

A mode is a class that runs some initialization code and some destruction code.
A mode has only one mandatory slot: the buffer it is associated to.
A common of use modes is to hold _key schemes_.

- Key schemes

A key scheme is roughly a plist of keymaps, e.g.

```
(:emacs my-emacs-map
 :vi-normal my-vi-map)
```

- Logic

When we press a key, Next will go through the active modes of the current buffer and fetch the keymap corresponding the `current-key-scheme` until it hits a binding.

This makes it possible to change binding scheme at run time, for individual buffers (or globally).

From there, VI implementation is trivial: the "insert mode" is just one binding scheme where `ESCAPE` is bound to `vi-normal-mode` and vice-versa, with `i` being bound to `vi-insert-mode` in vi-insert-mode.

This does not cover the advanced features mentioned in #28, such as
- Automatic resolution of conflicting key bindings with a timeout
- Command repeating
- Numeric argument
- Macros

All this will have to wait for some time :)

I'm planning to merge on Monday.
Thoughts? @jmercouris @vindarel @wasamasa @lehmacdj @jgkamat @noctuid

**Comment by Ambrevar** [comment]
Update:
- VI can now be set to be the default with
```
(add-to-default-list 'vi-normal-mode 'buffer 'default-modes)
```

- VI mode can now be properly de-activated.

- Shifted keys are properly translated, e.g. "s-o" is translated to "O" and "s-:" is translated to ":".

**Comment by jmercouris** [comment]
Everything looks OK to me so far. Will the missing features be implemented in a separate branch?

**Comment by Ambrevar** [comment]
Incoming just now!

File: source/keymap.lisp [hunk 1]
```diff
@@ -1,27 +1,59 @@
...
+                         for fun = (gethash key keymap)
+                         unless fun
+                           do (log:debug "Key-chord ~a normalized to ~a" key key-normal)
+                              (setf fun (gethash key-normal keymap))
+                         when fun
+                           return (cons fun mode))))
+    (when fun+mode
+      (if (eq (first fun+mode) #'prefix)
+          (first fun+mode)
+          (lambda () (funcall (first fun+mode) (cdr fun+mode)))))))
```
Comment (vindarel) [inline source/keymap.lisp]: nitpicking: mixed "first" and "cdr".

File: source/keymap.lisp [hunk 2]
```diff
@@ -42,103 +74,60 @@ For example, it may add C-M-s or C-x to a stack which will be consumed by
...
-         (key-maps (list (root-mode-default-keymap) local-map))
-         (serialized-key-stack (mapcar #'serialize-key-chord (key-chord-stack *interface*))))
-    (dolist (map key-maps)
-      (let ((bound-function (gethash serialized-key-stack map)))
-        (cond ((equalp "prefix" bound-function)
-               (return-from consume-key-sequence t))
-              (bound-function
+          ((minibuffer-active active-window)
+           (if (member "R" (key-chord-modifiers (first (key-chord-stack *interface*)))
+                       :test #'string=)
```
Comment (vindarel) [inline source/keymap.lisp]: Maybe this pattern can be extracted, I saw it twice or more (`"member … (key-chord-modifiers (first …) :test #'string=`).

File: source/mode.lisp [hunk 3]
```diff
@@ -7,34 +7,72 @@
...
-
-(defun root-mode-default-keymap ()
-  "Return the default keymap of root mode."
-  (eval (closer-mop:slot-definition-initform
-         (find-slot (find-class 'root-mode) 'keymap))))
+     (buffer :accessor buffer :initarg :buffer)
+     (activate :accessor activate :initarg :activate) ; TODO: This can be used in the future to temporarily turn off modes without destroying the object.
+     (destructor :accessor destructor  ; TODO: Better name?
+                 :documentation
+                 "A lambda function which takes the mode as argument.")
```
Comment (vindarel) [inline source/mode.lisp]: better docstring for destructor and docstrings for other slots ?

**Comment by vindarel** [comment]
\+1. Don't forget the user and developer documentation though. You already explained quite a bit in your above comment.

File: source/keymap.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/keymap.lisp]: Do you know the equivalent of `first` for `cdr`?
Or do you mean I should use `car` and `cdr` here?

File: source/keymap.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/keymap.lisp]: Yes, this `(member ... :test #'string=) is a bit verbose.  Isn't there some higher-level function in the standard already?  Maybe in the library?

File: source/keymap.lisp (hunk 1 above)
Comment (vindarel) [reply source/keymap.lisp]: yes `cdr` == `rest`.

(https://lispcookbook.github.io/cl-cookbook/data-structures.html#carcdr-or-firstrest-and-second-to-tenth ;) )

Comment (Ambrevar) [reply source/keymap.lisp]: Oops!  Must have been tired on this one... :)

File: source/minibuffer.lisp [hunk 4]
```diff
@@ -386,26 +386,42 @@
...
+                          (body :border-top "4px solid dimgray"
+                                :margin "0"
+                                :padding "0 6px")
+                          (p :margin "0")))))
+            (set-input minibuffer
+                       (cl-markup:markup
+                        (:head (:style style))
+                        (:body
+                         (:p (apply #'format nil args)))))))
+        (log:warn "Can't echo ~a without minibuffer or interface" args))))
```
Comment (jmercouris) [inline source/minibuffer.lisp]: Possibly also echo what it was trying to echo to stdout here?

Comment (Ambrevar) [reply source/minibuffer.lisp]: Hmm?  Isn't it exactly what `log:warn` does?

**Comment by noctuid** [comment]
I think the idea of key schemes is much better than trying to imitate what evil does. Can only one key scheme be active at a time? It seems like it would make sense to be able to have inheritance in some cases like evil does.

Is there currently some hook for when a new url is opened that could be used to set up local modes (e.g. I create a mode for reddit and enable it if the url matches reddit.com)?

**Comment by Ambrevar** [comment]
Only one key scheme is active at a time for a buffer.
By inheritence, you mean inheritance across schemes?

For instance if C-x C-c is defined in the `:emacs` scheme but not in VI, you'd like that it still works when the VI scheme is selected?

I'm not too sure about that, since it could be confusing to the user.  If we enable inheritance, can the user disable it?  How?  Can it be disable on a per-mode / per-buffer basis?

Hooks: while it's possible to program something, a convenient interface for hooking is missing at the moment.  I'll work on it soon, probably when I get down to adblocking.

**Comment by vindarel** [comment]
> a convenient interface for hooking is missing at the moment.

We can use CLOS's before/after methods for our classes, but it's like we had one single hook.

(https://lispcookbook.github.io/cl-cookbook/clos.html#method-qualifiers-before-after-around)

**Comment by jmercouris** [comment]
My vote is to restore the old hook logic, I think it was pretty simple/straightforward.

**Comment by noctuid** [comment]
> By inheritence, you mean inheritance across schemes?

Yeah, across schemes.

> For instance if C-x C-c is defined in the :emacs scheme but not in VI, you'd like that it still works when the VI scheme is selected?

That could be something someone would want, so I think it would make sense to allow that, but I was thinking more like evil's inheritance between states. Some keybindings will potentially make sense to have in multiple key schemes (e.g. normal and with a visual selection; probably a lot less keybindings will make sense in multiple key schemes than do for evil). You could duplicate them or have a variable to hold common keybindings, but I think it would make more sense to allow the user to change keybindings for both in the same place using `define-key`. There would be a key scheme whose only purpose was to hold common keybindings. It could be simpler to just bind in multiple keyschemes at once, but this seems a little less clean to me and could also potentially cause confusion. It's something worth considering I think.

> If we enable inheritance, can the user disable it?  How? Can it be disable on a per-mode / per-buffer basis?

Make the current active schemes optionally a list. It being global would be fine for the use case I had in mind, but having a buffer-local `current-key-scheme` might be useful in some cases. By default, I wouldn't do anything like inherit `:emacs` in a vi scheme.

**Comment by Ambrevar** [comment]
@jmercouris: The old hook system had a few issues, among which that it only ran at a predefined point in a command.  So it would not run on class initialization for instance, and it would not run before, after or around like is customary for auxiliary methods (or Emacs `advice` system).

I suggest we keep thinking about it for a couple of days until we find something that flexible enough: in the end it's a feature that is going to be exposed to the user and that will be key to customization.  For everything.

**Comment by Ambrevar** [comment]
@noctuid: Lots of good points here.  Merging schemes would be easy, but then you lose the ability to set a key in one place.  Hmm... Need to think a bit more about it.  Anyways, I'll probably merge this and keep inheritance for later.

**Comment by jmercouris** [comment]
Merge it! I’m excited to see it on master!

**Pull request merged**